src/tag.rs

```diff
@@ -58,183 +81,231 @@ impl Tag {
+    pub const CTX_2: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x02, 0, 0, 0]);
 
     /// The tag context specific tag [3].
-    pub const CTX_3: Self = Tag(Tag::CONTEXT_SPECIFIC | 3);
+    pub const CTX_3: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x03, 0, 0, 0]);
 
     /// The tag context specific tag [4].
-    pub const CTX_4: Self = Tag(Tag::CONTEXT_SPECIFIC | 4);
+    pub const CTX_4: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x04, 0, 0, 0]);
 
     /// The tag context specific tag [5].
-    pub const CTX_5: Self = Tag(Tag::CONTEXT_SPECIFIC | 5);
+    pub const CTX_5: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x05, 0, 0, 0]);
 
     /// The tag context specific tag [6].
-    pub const CTX_6: Self = Tag(Tag::CONTEXT_SPECIFIC | 6);
+    pub const CTX_6: Self = Tag([Tag::CONTEXT_SPECIFIC | 0x06, 0, 0, 0]);
 }
 
 impl Tag {
+    /// Encodes a number into the identifier representation. There are two forms:
+    /// * low tag number (for tag numbers between 0 and 30):
+    ///     One octet. Bits 8 and 7 specify the class , bit 6 indicates whether the encoding
+    ///     is primitive (0), and bits 5-1 give the tag number.
+    /// * High tag number (for tag numbers 31 and greater):
+    ///     Two or more octets. First octet is as in low-tag-number form, except that bits 5-1 all
+    ///     have value 1.
+    ///     Second and following octets give the tag number, base 128, most significant digit first,
+    ///     with as few digits as possible, and with the bit 8 of each octet except the last set
+    ///     to 1.
+    //
+    /// # Panics
+    ///
+    /// Currently, this function panics if the tag number is greater than
+    /// `MAX_VAL_SPAN_3_OCTETS`.
+    #[inline]
+    fn new(number: u32, type_mask: u8) -> Self {
+        assert!(number <= Tag::MAX_VAL_SPAN_3_OCTETS);
+        if number <= Tag::MAX_VAL_FOURTH_OCTET {
+            Tag([type_mask | number as u8, 0, 0, 0])
+        } else if number <= Tag::MAX_VAL_SPAN_1_OCTET {
+            // Fit the number in the third octets
+            let number = number as u8;
+            Tag([type_mask | Tag::SINGLEBYTE_DATA_MASK, number, 0, 0])
+        } else if number <= Tag::MAX_VAL_SPAN_2_OCTETS {
+            // Fit the number in the second and the third octets
+            let first_part = Tag::MULTIBYTE_DATA_MASK & ((number >> 7) as u8) | Tag::LAST_OCTET_MASK;
+            let second_part = Tag::MULTIBYTE_DATA_MASK & (number as u8);
+            Tag([type_mask | Tag::SINGLEBYTE_DATA_MASK, first_part, second_part, 0])
+        } else {
+            // Fit the number in the first, second and the third octets
+            let first_part = Tag::MULTIBYTE_DATA_MASK & ((number >> 14) as u8) | Tag::LAST_OCTET_MASK;
+            let second_part = Tag::MULTIBYTE_DATA_MASK & ((number >> 7) as u8) | Tag::LAST_OCTET_MASK;
+            let third_part = Tag::MULTIBYTE_DATA_MASK & (number as u8);
+            Tag([type_mask | Tag::SINGLEBYTE_DATA_MASK, first_part, second_part, third_part])
+        }
+    }
```

Comment:
Formatting nitpick: I like to have an empty line between methods.

Reply:
Fixed.